trace/trace.go

```diff
@@ -44,19 +68,61 @@ func writeFields(b *bytes.Buffer, l []Field) {
 	}
 }
 
+func writeMainLog(b *bytes.Buffer, msg string, totalTime time.Duration, startTime time.Time, fields []Field) {
+	b.WriteString(fmt.Sprintf("%q ", msg))
+	if len(fields) > 0 {
+		writeFields(b, fields)
+		b.WriteString(" ")
+	}
+
+	b.WriteString(fmt.Sprintf("%vms (%v)", durationToMilliseconds(totalTime), startTime.Format("15:04:00.000")))
+}
+
+func durationToMilliseconds(timeDuration time.Duration) int64 {
+	return timeDuration.Nanoseconds() / 1e6
+}
+
 type traceStep struct {
 	stepTime time.Time
 	msg      string
 	fields   []Field
 }
 
+func (s traceStep) time() time.Time {
+	return s.stepTime
+}
+
+func (s traceStep) writeStep(b *bytes.Buffer, formatter string, stepThreshold time.Duration,
+	lastStepTime time.Time) time.Time {
+	stepDuration := s.stepTime.Sub(lastStepTime)
+	if stepThreshold == 0 || stepDuration > stepThreshold || klog.V(4) {
+		b.WriteString(fmt.Sprintf("%v---", formatter))
+		writeMainLog(b, s.msg, stepDuration, s.stepTime, s.fields)
+	}
+	return s.stepTime
+}
+
 // Trace keeps track of a set of "steps" and allows us to log a specific
 // step if it took longer than its share of the total allowed time
 type Trace struct {
-	name      string
-	fields    []Field
-	startTime time.Time
-	steps     []traceStep
+	name        string
+	fields      []Field
+	startTime   time.Time
+	steps       []traceStep
+	nestedTrace []*Trace
```

Comment:
Any reason we need two separate arrays here, or would it work to just do `steps []stepTraceTime`?

Reply:
I think it is still logical to separate them. Although making it one array will remove the need to sort it. @sttts any thoughts?

Reply:
It’s fine to have a slice of mixed types. I don’t think the added complexity of keeping them separate is worth it.

Reply:
Okay. That's fine
The only consideration I have seen so far is in this [function](https://github.com/kubernetes/utils/blob/master/trace/trace.go#L123) where the number of steps is used to calculate the step threshold. Wdyt?

Reply:
This is a really good question. I think we need to be careful with thresholds and logging.

Just using the number of steps (incl nested traces) of the top level trace to calculate the step threshold like you linked to doesn't seem to make much sense. Each of the nested traces can contain many more steps. Using the current logic could easily result in nested trace steps getting unreasonable thresholds, resulting in the logs getting spammed.

There is a loosely related problem. Right now it looks like if we call `LogIfLong` on a nested trace, it logs it separate from the surrounding trace. So if we ask callers to use functions like `NewNestedInContext` (https://github.com/kubernetes/kubernetes/pull/87610/files#diff-be0415fae94fa4b972093826630556d9R104) where they don't know if they ended up creating a top level trace or a nested trace, we have a problem. The caller will typically `LogIfLong` on the trace, and as a result it get it logged twice, once when `LogIfLong` is called directly on the nested trace and again when it is called on the top level trace. The more level of nesting the worse this would be.

I think we need to keep track of the duration passed to `LogIfLong` on calls to nested traces, but don't log immediately and instead use it to calculate the thresholds. If we just ignored `LogIfLong` calls to nested traces we have to try to calculate a reasonable threshold for the nested trace and risk getting the logs spammed if we get the value wrong. Better take advantage of the hint the caller has provided.

Unfortunately, this is going to complicate the threshold calculations.

Straw-man proposal:

  - Find the nested traces that have a threshold provided to them by a call to LogIfLong
  - Subtract the nested trace thresholds from the threshold of the current trace to calculate a remaining threshold for the steps of the current trace (but always using at least some ratio of the current trace threshold when calculating the stepThreshold, e.g. no less than 1/4)
  - Divide the remaining threshold evenly among the steps of the trace like we do today
  - If a nested trace does not have it's own threshold provided by a call to LogIfLong then
    treat it as a step and give it an even division of the threshold
  - Don't pass stepThreshold into nested traces. Each nested trace calculates its own stepThreshold.

Suggestions on how to better do this welcome.

Reply:
I think the first hurdle is how to know if LogIfLong has been called on a trace. We could do this by adding an extra field in the struct `threshold` so LogIfLong sets this field and we can easily access it when logging the trace.

Also, I am not clear on how we want the trace to behave when LogIfLong has been called both on the parent trace and the nested trace. Do we want them to log separately or not at all?

Reply:
> I think the first hurdle is how to know if LogIfLong has been called on a trace. We could do this by adding an extra field in the struct `threshold` so LogIfLong sets this field and we can easily access it when logging the trace.

I think this is needed. It can be pointer. If the pointer is nil, we know LogIfLong has not been called.

> Also, I am not clear on how we want the trace to behave when LogIfLong has been called both on the parent trace and the nested trace. Do we want them to log separately or not at all?

I tried to answer this in my above comment. I think we should consider this approach:

- We only log when LogIfLong is called on the top level trace.
- Calls to LogIfLong on nested traces simply keep track of the threshold.
- When actually logging the trace (when LogIfLong is called on the top level trace), log any nested traces that exceed their threshold even if the traces they are nested in do not exceed their threshold.

The one downside to this approach is that if a nested trace exceeds its threshold but the trace it is nested under never calls LogIfLong, then the nested trace won't ever get logged. In practice LogIfLong should be called in a defer so this shouldn't be a problem, but it needs to be documented.